pact/Pact/Core/Legacy/LegacyCodec.hs

```diff
@@ -0,0 +1,163 @@
+        String n -> case readMaybe (unpack n) of
+          Just i -> return i
+          Nothing -> fail $ "Invalid integer value: " ++ show s
+        _ -> fail $ "Invalid integer value: " ++ show s
+    {-# INLINE decodeInteger #-}
+    aesonEncodeInteger i
+      | isSafeInteger i = pairs ( field .= i )
+      | otherwise = pairs ( field .= show i )
+    field = "int"
+
+-- | Decimals encode to a Scientific, which is encoded as an object + String
+-- if mantissa precision exceeds JS.
+-- TODO fromRational . toRational may not be the speediest.
+decimalCodec :: Codec Decimal
+decimalCodec = Codec enc dec encValue aesonEnc
+  where
+    enc d@(Decimal _places mantissa)
+      | isSafeInteger mantissa = J.build $ J.Aeson @Scientific $ fromRational $ toRational d
+      | otherwise = J.object [ "decimal" J..= pack (show d) ]
+    {-# INLINE enc #-}
+    encValue d@(Decimal _places mantissa)
+      | isSafeInteger mantissa = Number $ fromRational $ toRational d
+      | otherwise = object [ field .= show d ]
+    {-# INLINE encValue #-}
+    dec (Number n) = return $ fromRational $ toRational n
+    dec (A.Object o) = o .: field >>= \s -> case readMaybe (unpack s) of
+      Just d -> return d
+      Nothing -> fail $ "Invalid decimal value: " ++ show s
+    dec v = fail $ "Invalid decimal value: " ++ show v
+    {-# INLINE dec #-}
+    aesonEnc d@(Decimal _places mantissa)
+      | isSafeInteger mantissa = A.scientific $ fromRational $ toRational d
+      | otherwise = pairs ( field .= show d )
+    {-# INLINE aesonEnc #-}
+    field = "decimal"
+
+-- | default Pact ISO8601 format
+pactISO8601Format :: String
+pactISO8601Format = "%Y-%m-%dT%H:%M:%SZ"
+
+-- | high-precision format
+highPrecFormat :: String
+highPrecFormat = "%Y-%m-%dT%H:%M:%S.%vZ"
+
+-- | Time uses
+timeCodec :: Codec UTCTime
+timeCodec = Codec enc dec encValue aesonEnc
+  where
+    enc t
+      | 1 == denom t = J.object [ "time" J..= pack (formatTime pactISO8601Format t) ]
+      | otherwise = J.object [ "timep" J..= pack (formatTime highPrecFormat t) ]
+    {-# INLINE enc #-}
+    encValue t
+      | 1 == denom t = object [ field .= formatTime pactISO8601Format t ]
+      | otherwise = object [ highprec .= formatTime highPrecFormat t ]
+    {-# INLINE encValue #-}
+    dec = withObject "time" $ \o ->
+      (o .: field >>= mkTime pactISO8601Format) <|>
+      (o .: highprec >>= mkTime highPrecFormat)
+      where
+        mkTime :: String -> String -> Parser UTCTime
+        mkTime fmt v = case parseTime fmt v of
+              Just t -> return t
+              Nothing -> fail $ "Invalid time value, expected " ++ fmt
+    {-# INLINE dec #-}
+    aesonEnc t
+      | 1 == denom t = pairs ( field .= formatTime pactISO8601Format t )
+      | otherwise = pairs ( highprec .= formatTime highPrecFormat t )
+    {-# INLINE aesonEnc #-}
+    field = "time"
+    highprec = "timep"
+
+    -- FIXME this should be 1000000. Fixing this would break mainnet replay
```

Comment:
iirc, this `FIXME` is a copyover from prod?

Reply:
yes